### ISSUE TITLE
[Feature req] Change order of EPG items in iptv merge and add internal EPGs from 'Add EPG' submenu [source type]

### ISSUE THREAD

**Issue by enerdude** [matthuisman/slyguy.addons#894]
Leading on from #893 I have done some more testing....

With the old iptv merge I initially did not get any joy with my own EPG (containing duplicate as well as unique channel ids) in conjunction with the internal EPG - it seems I just had to clean the guide and channel cache in Kodi settings to get it working.

Further, on the old setup before you put an option to disable the internal EPG, the internal EPG got merged before my own EPG and all the channels that was duplicate would take preference from the internal EPG (it still added missing channels from my EPG though).  For my use case scenario this was actually what I wanted because the internal EPG has more channel details / thumbs / icons than my own EPG - I mainly use my own EPG to fill in the gaps for missing channels.

Should I be able to source my own "perfect" EPG I would just use the new iptv merge 0.27.3 to ignore the internal EPG completely.  Until then I want to combine it with my own EPG.

As you can see the order the EPGs gets merged in makes a difference, especially when duplicate channel ids is found.

I therefore request the following feature(s)

1.  Add a context menu to EPG items added to be able to move them up and down (like in playlists menu)
2.  When in EPGs --> Add EPG --> Select source type --> Add-on, I get an error "No compatible addons found" although e.g. Dstv now is selected in playlists.  Obviously I can add the source (https://i.mjh.nz/DStv/za.xml) manually but not everyone has this detail and it would be nice that the program can just add the internal EPG as a seperate EPG source that one can move up and down to suit your priorities of all your EPGs.

Thanks for all your time with these addons.
Regards,

**Comment by matthuisman** [comment]
i think a simpler fix is just making internal epgs merged last. ill push that change shortly

**Comment by matthuisman** [comment]
wait, it should already do them last...

Do you need internal merged first or last?

**Comment by matthuisman** [comment]
also, I've just pushed [plugin.video.dstv.now -> v0.18.1](https://github.com/matthuisman/slyguy.addons/commit/d2bac1501ba5844c52a3e86621a894615b6d9283) which moves back to "epg" separate from playlist.

That makes it / allows it to be added under EPGS seperately.
Itll automatically get added when you add the playlist source.

#2 i cant really do easily as the DB tables are already set.
But you can just delete source and re-add to change the order etc

**Comment by enerdude** [comment]
> wait, it should already do them last...
> 
> Do you need internal merged first or last?

At the moment I need it first cause the internal EPG has better details in the EPG items, so my own EPG merges second to just fill the gaps (channel ids not found in internal EPG, albeit with less info on the items)

Should [Tinuva](https://github.com/tinuva/) update the DSTV epg to have all channels with info, I will only use the internal EPG.  However if I can find the "perfect" EPG before he does, I can use my own EPG as the only EPG

On the other hand, there might be cases where I find an EPG with same detail level as the internal EPG but with less channel ids, where I would then need to merge my EPG first and then the internal EPG second to fill in the gaps.

So it all depends on the situation

**Comment by enerdude** [comment]
> also, I've just pushed [plugin.video.dstv.now -> v0.18.1](https://github.com/matthuisman/slyguy.addons/commit/d2bac1501ba5844c52a3e86621a894615b6d9283) which moves back to "epg" separate from playlist.

I am getting an error upon opening the v0.18.1:

`2024-10-08 22:05:58.999 T:32440   error <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'ImportError'>
                                                   Error Contents: cannot import name 'ZA_EPG_URL' from 'resources.lib.constants' (C:\Users\Peter\AppData\Roaming\Kodi\addons\plugin.video.dstv.now\resources\lib\constants.py)
                                                   Traceback (most recent call last):
                                                     File "C:\Users\Peter\AppData\Roaming\Kodi\addons\plugin.video.dstv.now\default.py", line 1, in <module>
                                                       from resources.lib.plugin import plugin
                                                     File "C:\Users\Peter\AppData\Roaming\Kodi\addons\plugin.video.dstv.now\resources\lib\plugin.py", line 14, in <module>
                                                       from .constants import ZA_EPG_URL, LICENSE_COOLDOWN
                                                   ImportError: cannot import name 'ZA_EPG_URL' from 'resources.lib.constants' (C:\Users\Peter\AppData\Roaming\Kodi\addons\plugin.video.dstv.now\resources\lib\constants.py)
                                                   -->End of Python script error report<--
                                                   
2024-10-08 22:05:59.278 T:20816   error <general>: XFILE::CDirectory::GetDirectory - Error getting plugin://plugin.video.dstv.now/
2024-10-08 22:05:59.280 T:21496   error <general>: CGUIMediaWindow::GetDirectory(plugin://plugin.video.dstv.now/) failed`

> That makes it / allows it to be added under EPGS seperately. Itll automatically get added when you add the playlist source.
> 
> #2 i cant really do easily as the DB tables are already set. But you can just delete source and re-add to change the order etc

Ok its a pity but like you say one can just delete all playlist and add them in your preferred order

BTW, thanks for your quick responses, did not expect you to improve the add-on so quick

**Comment by matthuisman** [comment]
oops, fixed that in 0.18.2 :)

**Comment by enerdude** [comment]
For some odd reason I can get no EPG info to show up in pvr simple client.
I first went to
Kodi settings --> PVR & Live TV --> Guide --> Clear data, then
Kodi settings --> PVR & Live TV --> General --> Clear data
to start afresh

I reinstalled iptv merge and iptv simple client but it still does not want to show any info
I have checked userdata\addon_data\plugin.program.iptv.merge\ and both the playlist.m3u8 and epg.xml is there, both with valid data.
I tried without extra playlists and epgs, just using dstv playlist and dstv EPG.


Can you please check from your side?
Thanks in advance

**Comment by matthuisman** [comment]
i have nothing to check my side? maybe check the epg.xml actually has data from today etc etc

**Comment by enerdude** [comment]
Ok did some investigating.  I am running the latest official release of Kodi (21.1) on a windows and android system.
The version of pvr.iptvsimple I have is 21.9.2.  Seems there is some kind of problem with this version.  I tried everything and could not get guide data to show (the epg.xml is fine and has current info)

So I uninstalled my Kodi and installed a fresh one without any profile (keeping a seperate backup of my profile)
First thing I noticed when re-installing dstv and iptv merge was that iptv merge gave me an error ("pvr.iptvsimple add-on is required") when trying to install iptv simple client from the iptv merge menu.  I then checked the kodi repository and realised kodi 21.1 does not come with iptv simple client by default (weird!)

I installed pvr.iptvsimple 21.8.3 (mentioned [here](https://kodi.tv/addons/omega/pvr.iptvsimple/) on Kodi's official download site)
The ftp download link they give there was dead so I had to google for 21.8.3 zip on google.  Installed 21.8.3 and configured iptv merge with just dstv playlist and epg and it worked first time.  I then saw that under addons their was a update waiting for iptv simple client 21.9.2 but I set iptv simple client to not update automatically as this would cause my problem again (I did update it just to confirm that the error came back and then downgraded to 21.8.3 again and it works fine).

Finally I deleted the kodi profile and copied my old profile back, downgraded to 21.8.3, cleared guide and playlist data in kodi settings, ran iptv merge again and everything is now working fine.
